include/eve/detail/function/simd/x86/swizzle.hpp

```diff
@@ -0,0 +1,258 @@
+    for(int i=c/2;i<c;++i)  if(idx[i] <  c/2 && idx[i] != -1) return false;
+
+    return true;
+  }
+
+  //================================================================================================
+  // SSE2-SSSE3 variant
+  //================================================================================================
+  template<typename T, typename N, x86_abi ABI, shuffle_pattern Pattern>
+  EVE_FORCEINLINE auto basic_swizzle_( EVE_SUPPORTS(sse2_), wide<T,N,ABI> const& v, Pattern const&)
+  {
+    constexpr auto sz = Pattern::size(N::value);
+    using that_t      = as_wide_t<wide<T,N,ABI>,fixed<sz>>;
+
+    constexpr Pattern q = {};
+
+    // We're swizzling so much we aggregate the output
+    if constexpr( has_aggregated_abi_v<that_t> )
+    {
+      return aggregate_swizzle(v,q);
+    }
+    else if constexpr( current_api >= ssse3 )
+    {
+      using st_t    = typename that_t::storage_type;
+      using bytes_t = typename that_t::template rebind<std::uint8_t,fixed<16>>;
+      using i_t     = as_integer_t<wide<T,N,ABI>>;
+
+      return that_t ( (st_t)_mm_shuffle_epi8( bit_cast(v,as_<i_t>{}).storage()
+                                            , as_bytes<that_t>(q,as_<bytes_t>())
+                                            )
+                    );
+    }
+    else
+    {
+      if constexpr(sizeof(T) == 8)
+      {
+        using f_t         = as_floating_point_t<that_t>;
+        auto const vv     = bit_cast(v,as_<f_t>{});
+        constexpr auto m  = _MM_SHUFFLE2(q(1,2)&1, q(0,2)&1);
+        return bit_cast(process_zeros(f_t{_mm_shuffle_pd(vv,vv,m)},q),as_<that_t>{});
+      }
+      else if constexpr(sizeof(T) == 4)
+      {
+        constexpr auto m  = _MM_SHUFFLE(q(3,4)&3, q(2,4)&3, q(1,4)&3, q(0,4)&3);
+        if constexpr( std::same_as<T,float> ) return process_zeros(that_t{_mm_shuffle_ps(v,v,m)} ,q);
+        else                                  return process_zeros(that_t{_mm_shuffle_epi32(v,m)},q);
+      }
+      else if constexpr(sizeof(T) == 2)
+      {
+        if constexpr( (sz < 8) && (q < 4) )
+        {
+          constexpr auto m  = _MM_SHUFFLE(q(3,4)&3, q(2,4)&3, q(1,4)&3, q(0,4)&3);
+          return process_zeros(that_t{_mm_shufflelo_epi16(v,m)},q);
+        }
+        else if constexpr( (sz < 8) && (q >= 4) )
+        {
+          constexpr auto m  = _MM_SHUFFLE(q(3,4)&3, q(2,4)&3, q(1,4)&3, q(0,4)&3);
+          return process_zeros(that_t{_mm_bsrli_si128(_mm_shufflehi_epi16(v,m),8)},q);
+        }
+        else
+        {
+          constexpr auto lp = pattern_view<0,4,8>(q);
+          constexpr auto hp = pattern_view<4,8,8>(q);
+
+                if constexpr( lp < 4 && hp >= 4) return process_zeros(that_t{v[lp],v[hp]},q);
+          else  if constexpr( hp < 4 && lp >= 4) return process_zeros(that_t{v[lp],v[hp]},q);
+          else  return basic_swizzle_(EVE_RETARGET(cpu_),v,q);
+        }
+      }
+      else
+      {
+        return basic_swizzle_(EVE_RETARGET(cpu_),v,q);
+      }
```

Comment:
Chars should convert to shorts, swizzle and convert back I believe.

Reply:
it won't help the case like pattern<1,0,3,2,...> where you need to shuffle at the byte level.
We can add those special cases later IMHO. The pattern to detect is non-trivial and will probably make the PR larger than required.

Reply:
? It will. You just convert to wide<16> shorts from wide<16> chars. We will do that later.